watertap/property_models/anaerobic_digestion/adm1_reactions.py

```diff
@@ -0,0 +1,1433 @@
+            ("R19", "Liq", "X_I"): 0,
+        }
+
+        # TODO: Add S_IC stoichiometric coefficients that depend on other stoich coefficients (R5, 6, 10, 11, 12)
+        #  - determine conversion factors for S_IC stoichiometry as/if needed
+        # s_ic_rxns = ["R5", "R6", "R10", "R11", "R12"]
+        s_ic_rxns = self.rate_reaction_idx
+        for R in s_ic_rxns:
+            self.rate_reaction_stoichiometry[R, "Liq", "S_IC"] = -sum(
+                self.Ci[S] * self.rate_reaction_stoichiometry[R, "Liq", S]
+                for S in Ci_dict.keys()
+                if S != "S_IC"
+            )
+
+        # Fix all the variables we just created
+        for v in self.component_objects(pyo.Var, descend_into=False):
+            v.fix()
+
+    @classmethod
+    def define_metadata(cls, obj):
+        obj.add_properties(
+            {"reaction_rate": {"method": "_rxn_rate"}, "I": {"method": "_I"}}
+        )
+        obj.add_default_units(
+            {
+                "time": pyo.units.s,
+                "length": pyo.units.m,
+                "mass": pyo.units.kg,
+                "amount": pyo.units.kmol,
+                "temperature": pyo.units.K,
+            }
+        )
+
+
+class _ADM1ReactionBlock(ReactionBlockBase):
+    """
+    This Class contains methods which should be applied to Reaction Blocks as a
+    whole, rather than individual elements of indexed Reaction Blocks.
+    """
+
+    def initialize(blk, outlvl=idaeslog.NOTSET, **kwargs):
+        """
+        Initialization routine for reaction package.
+
+        Keyword Arguments:
+            outlvl : sets output level of initialization routine
+
+        Returns:
+            None
+        """
+        init_log = idaeslog.getInitLogger(blk.name, outlvl, tag="properties")
+        init_log.info("Initialization Complete.")
+
+
+@declare_process_block_class("ADM1ReactionBlock", block_class=_ADM1ReactionBlock)
+class ADM1ReactionBlockData(ReactionBlockDataBase):
+    """
+    ReactionBlock for ADM1.
+    """
+
+    def build(self):
+        """
+        Callable method for Block construction
+        """
+        super().build()
+
+        # Create references to state vars
+        # Concentration
+        add_object_reference(self, "conc_mass_comp_ref", self.state_ref.conc_mass_comp)
+
+    def _pH(self):
+        def rule_pH(b):
+            # TODO: determine how to handle hydrogen ion concentration - add as Solute or include as variable/expression?
+            # For now, use self.S_H
+            return -pyo.log10(b.S_H)
+
+        self.pH = pyo.Expression(rule=rule_pH, doc="pH of solution")
+
+    def _I_IN_lim(self):
+        def rule_I_IN_lim(b):
+            # TODO: revisit molar concentration of S_IN as it is not in property model as of yet
+            return 1 / (
+                1 + b.params.K_S_IN / b.state_ref.conc_mol_phase_comp["Liq", "S_IN"]
+            )
+
+        self.I_IN_lim = pyo.Expression(rule=rule_I_IN_lim)
```

Comment:
Some doc strings on these might help users (and readers of the code like me) understand what these terms are.

Reply:
Will do.